### PR TITLE
Add Psalm

### PR DESCRIPTION
**Pull request by prolic**
resolves https://github.com/prooph/event-store-client/issues/127

depends on https://github.com/prooph/event-store/pull/398

psalm up to level 2

### CODE REVIEW

File: examples/demo-cluster.php [hunk 1]
```diff
@@ -60,6 +60,7 @@
         true
     );
 
+    /** @psalm-suppress ForbiddenCode */
```
Comment (enumag) [inline examples/demo-cluster.php:63]: I dislike having such annotations in the code. I suggest suppressing this rule for the examples directory in psalm configuration instead.

File: src/Messages/ClusterMessages/ClusterInfoDto.php [hunk 2]
```diff
@@ -17,10 +17,10 @@
 
 class ClusterInfoDto
 {
-    /** @var MemberInfoDto[] */
+    /** @var array<int, MemberInfoDto> */
```
Comment (enumag) [inline src/Messages/ClusterMessages/ClusterInfoDto.php:20]: Use list instead. https://psalm.dev/docs/annotating_code/type_syntax/array_types/#lists

File: examples/demo-cluster.php (hunk 1 above)
Comment (prolic) [reply examples/demo-cluster.php:63]: ok

File: src/Messages/ClusterMessages/ClusterInfoDto.php (hunk 2 above)
Comment (prolic) [reply src/Messages/ClusterMessages/ClusterInfoDto.php:20]: thanks, I didn't know

File: src/Internal/EventStoreConnectionLogicHandler.php [hunk 3]
```diff
@@ -344,6 +345,7 @@ public function tcpConnectionError(TcpPackageConnection $tcpPackageConnection, T
             return;
         }
 
+        /** @psalm-suppress PossiblyNullReference */
```
Comment (prolic) [inline src/Internal/EventStoreConnectionLogicHandler.php:348]: I had to add a bunch of those comments in this class. The class itself is internal and there is no way that the connection is null in those cases, but psalm can't know.

File: src/Internal/EventStoreConnectionLogicHandler.php [hunk 4]
```diff
@@ -657,7 +667,11 @@ private function startSubscription(StartSubscriptionMessage $message): void
                     $message->userCredentials(),
                     fn (EventStoreSubscription $subscription, ResolvedEvent $resolvedEvent): Promise => ($message->eventAppeared())($subscription, $resolvedEvent),
                     function (EventStoreSubscription $subscription, SubscriptionDropReason $reason, ?Throwable $exception = null) use ($message): void {
-                        ($message->subscriptionDropped())($subscription, $reason, $exception);
+                        $subscriptionDropped = $message->subscriptionDropped();
+
+                        if (null !== $subscriptionDropped) {
+                            $subscriptionDropped($subscription, $reason, $exception);
+                        }
```
Comment (prolic) [inline src/Internal/EventStoreConnectionLogicHandler.php:674]: I just catched this

File: src/Internal/EventStoreNodeConnection.php [hunk 5]
```diff
@@ -581,6 +586,7 @@ public function createPersistentSubscriptionAsync(
         return $deferred->promise();
     }
 
+    /** {@inheritdoc} */
```
Comment (prolic) [inline src/Internal/EventStoreNodeConnection.php:589]: inheritdoc or copy paste the phpdoc block?

**Comment by prolic** [comment]
psalm level 3 almost done, hopefully I can do the rest tomorrow and then take care of your unreference watcher pull request.

File: src/ClientOperations/AbstractOperation.php [hunk 6]
```diff
@@ -93,6 +104,7 @@ public function createNetworkPackage(string $correlationId): TcpPackage
     public function inspectPackage(TcpPackage $package): InspectionResult
     {
         if ($package->command()->equals($this->responseCommand)) {
+            /** @psalm-suppress InvalidStringClass */
             $responseMessage = new $this->responseClassName();
```
Comment (enumag) [inline src/ClientOperations/AbstractOperation.php:108]: The `responseClassName()` method should have a `@return class-string<...>` https://psalm.dev/docs/annotating_code/type_syntax/scalar_types/

File: src/ClientOperations/AbstractOperation.php [hunk 7]
```diff
@@ -60,9 +64,16 @@ public function __construct(
 
     abstract protected function createRequestDto(): Message;
 
+    /**
+     * @param TResponse $response
+     * @return InspectionResult
```
Comment (enumag) [inline src/ClientOperations/AbstractOperation.php:69]: Redunant annotation, remove.

File: src/ClientOperations/AbstractSubscriptionOperation.php [hunk 8]
```diff
@@ -337,6 +337,7 @@ protected function eventAppeared(ResolvedEvent $e): void
         }
 
         if ($this->verboseLogging) {
+            /** @psalm-suppress PossiblyNullReference */
```
Comment (enumag) [inline src/ClientOperations/AbstractSubscriptionOperation.php:340]: What is possibly null according to psalm here? And why are you sure it won't be null?

File: src/ClientOperations/AbstractOperation.php [hunk 9]
```diff
@@ -32,7 +32,11 @@
 use Psr\Log\LoggerInterface as Logger;
 use Throwable;
 
-/** @internal */
+/**
+ * @internal
+ * @template TResponse
```
Comment (enumag) [inline src/ClientOperations/AbstractOperation.php:37]: @template TResponse of Message

File: src/ClientOperations/AppendToStreamOperation.php [hunk 10]
```diff
@@ -83,10 +86,12 @@ protected function createRequestDto(): Message
         return $message;
     }
 
+    /**
+     * @param WriteEventsCompleted $response
+     * @return InspectionResult
```
Comment (enumag) [inline src/ClientOperations/AppendToStreamOperation.php:91]: reduntant return annotation

File: src/Internal/EventStoreNodeConnection.php (hunk 5 above)
Comment (enumag) [reply src/Internal/EventStoreNodeConnection.php:589]: I think the tools are using inheritdoc implicitly now so nothing should be necessary... but I might be wrong.

**Comment by prolic** [comment]
@enumag for psalm level 1 I receive lots of those errors:

```
ERROR: MixedReturnTypeCoercion - src/Projections/QueryManager.php:96:17 - The declared return type 'Amp\Promise<void>' for Prooph\EventStoreClient\Projections\QueryManager::waitForCompletedAsync is more specific than the inferred return type 'Amp\Promise<mixed>' (see https://psalm.dev/197)
    /** @return Promise<void> */
```

any ideas how I can bring psalm to understand this correctly without suppressing everything?

File: composer.json [hunk 11]
```diff
@@ -21,23 +21,24 @@
...
+    "amphp/http-client": "^4.2.2",
+    "amphp/byte-stream": "^1.6.3",
     "amphp/file": "^1.0.0",
-    "amphp/socket": "^1.0.0",
-    "google/protobuf": "^3.9",
-    "prooph/event-store": "dev-master",
-    "psr/log": "^1.1.0"
+    "amphp/socket": "^1.1.1",
+    "google/protobuf": "^3.11.4",
+    "prooph/event-store": "dev-psalm",
```
Comment (prolic) [inline composer.json:30]: todo: set back to `dev-master` once https://github.com/prooph/event-store/pull/398 is merged.

**Comment by prolic** [comment]
@enumag I try to work on your suggested comments tomorrow.

**Comment by enumag** [comment]
https://github.com/prooph/event-store-client/pull/137#issuecomment-625610105 - this seems like a psalm bug to me so suppress and report is the way to go imo

**Comment by prolic** [comment]
> [#137 (comment)](https://github.com/prooph/event-store-client/pull/137#issuecomment-625610105) - this seems like a psalm bug to me so suppress and report is the way to go imo

I cannot report a psalm bug, because the author hates me and banned me on github. Can you do it?

File: src/ClientOperations/AbstractSubscriptionOperation.php (hunk 8 above)
Comment (prolic) [reply src/ClientOperations/AbstractSubscriptionOperation.php:340]: `$e->originalEvent` could be null, but an event appeared, so it's not.

File: src/ClientOperations/AppendToStreamOperation.php (hunk 10 above)
Comment (prolic) [reply src/ClientOperations/AppendToStreamOperation.php]: done

File: src/ClientOperations/AbstractOperation.php (hunk 9 above)
Comment (prolic) [reply src/ClientOperations/AbstractOperation.php]: done

File: src/ClientOperations/AbstractOperation.php (hunk 7 above)
Comment (prolic) [reply src/ClientOperations/AbstractOperation.php]: done

File: src/ClientOperations/AbstractOperation.php (hunk 6 above)
Comment (prolic) [reply src/ClientOperations/AbstractOperation.php]: done

**Comment by prolic** [comment]
I'm going to merge this now. Further improvements can be done in a separate pull request. Psalm is now at level 2, level 1 can only be resolved once psalm-devs fix their problems.

**Pull request merged**